lib/services/workflow-api-service.js

```diff
@@ -128,7 +128,56 @@ function workflowApiServiceFactory(
     };
 
     WorkflowApiService.prototype.cancelTaskGraph = function(graphId) {
-        return taskGraphProtocol.cancelTaskGraph(graphId);
+        var instanceId;
+
+        return waterline.graphobjects.needByIdentifier(graphId)
+        .then(function(workflow) {
+            instanceId = workflow.instanceId;
+
+            if (workflow.active()) {
+                return taskGraphProtocol.cancelTaskGraph(instanceId);
+            }
+            throw new Errors.TaskCancellationError(
+                graphId + ' is not an active workflow'
+            );
+        })
+        .then(function(cancelledGraphId) {
+            if (cancelledGraphId !== instanceId) {
+                // Cancelled graph ID doesn't match the requested ID.
+                throw new Errors.TaskCancellationError('Failed to cancel ' + graphId);
+            }
+            return waterline.graphobjects.needByIdentifier(graphId);
+        });
+    };
+
+    WorkflowApiService.prototype.deleteTaskGraph = function(graphId) {
+        var instanceId;
+
+        // Taskgraph deletion sequence:
+        // 1) Get the graph object by ID
+        // 2) Check if the returned workflow is running.
+        // 3) If it is running, cancel it otherwise go on to step 4.
+        // 4) Delete the graph object from the task graph store.
+        return waterline.graphobjects.needByIdentifier(graphId)
+        .then(function(workflow) {
+            // Since deleteGraph takes instanceId as an argument, save it
+            // here using a closure so we don't need to do another query later.
+            instanceId = workflow.instanceId;
+            if (workflow.active()) {
+                return taskGraphProtocol.cancelTaskGraph(instanceId);
+            }
+            return instanceId;
+        })
+        .then(function(cancelledGraphId) {
+            if (cancelledGraphId !== instanceId) {
```

Comment:
I think this judgment is redundant, base on your design, the `cancelledGraphId` and `instanceId` is always the equivalent. So the next throwing will never be executed and this is always a coverage hole for unit-testing.